### ISSUE TITLE
Area2D mouse_entered and mouse_exited signals not called in Viewport

### ISSUE THREAD

**Issue by BeayemX** [godotengine/godot#26181]
**Godot version:**

Tested with 4f697f73a5ac567c1dc22ce44a7fab98f619dfe3

**OS/device including version:**

Linux Mint 19.1 tessa 64 Bit

**Issue description:**

When using Area2D in a Viewport the mouse_entered and mouse_exited signals are never called.

**Steps to reproduce:**
Use the test project.
When you run the project you see two godot-icons. Both icons are unmodified instances of `Area2D.tscn`.
The one on the left is contained in a Viewport and does not print any messages when it should receive mouse enter and exit signals.
The one on the right will print messages when the mouse enters/exits the icon.

Additional info: 
The UI node's mouse filter is set to `pass`.
The ViewportContainer node's mouse filter is set to `pass`.
The Viewport's setting `Handle input locally` was tested enabled and disabled.
The Viewport's setting `Disable input` is disabled.
The Viewport's setting `Own World` was tested enabled and disabled.

**Minimal reproduction project:**

[ViewportIssue.zip](https://github.com/godotengine/godot/files/2896342/ViewportIssue.zip)

**Comment by BeayemX** [comment]
It seems that all signals from `CollisionObject2D` are affected, because the signal `input_event` is also not called in the viewport instance.

**Comment by RandomShaper** [comment]
That is expected behavior. Picking works with unhandled events and currently `ViewportContainer` pass to its `Viewport` as unhandled events only those really unhandled in the whole scene.

You need something like this in your `Viewport`:
```
extends Viewport

func _input(event):
	unhandled_input(event)
```

I'll send a PR that allows to have that working automatically, but I don't think it will do it into 3.1.0.

**Comment by cobreen** [comment]
I'm using the latest version. It's still not working

**Comment by giulianob** [comment]
@RandomShaper can we reopen the issue to track it? I just ran into the same issue and it doesnt seem obvious as to the end user why this would not work.

**Comment by Skaruts** [comment]
@giulianob indeed. It took me a lot of asking around for a problem that had no obvious solution, until I figured this ought to be an issue. In my case it involves 3D Areas. 

Thanks for that workaround @RandomShaper.

**Comment by Zylann** [comment]
> Picking works with unhandled events and currently ViewportContainer pass to its Viewport as unhandled events only those really unhandled in the whole scene.
@RandomShaper I don't understand, in your example code you used `_input`, but this is not unhandled events, it's going to catch them all, isn't it?
Also, if I click an Area2D inside a viewport, and the area doesn't react, who is actually handling that input? And how to make sure it gets forwarded to that viewport and not to some other random script in the game that might have `_unhandled_input`? Because if input is unhandled *so far* in the enclosing scene, it might become handled after being checked by objects inside the viewport.

**Comment by Skaruts** [comment]
@Zylann , I did have to turn input on/off manually for the viewport, using `mouse_entered/exited` signals on the parent control, or else it would still receive it while the mouse was over the UI around it.  

I suppose this is also related to your other concern, since the input I'm using in the scene that's under the viewport is actually Area input_event signals, for mouse clicks and movement, which according to the docs are not unhandled_input and would normally be handled after that.

**Comment by cjhorton** [comment]
> @Zylann , I did have to turn input on/off manually for the viewport, using `mouse_entered/exited` signals on the parent control, or else it would still receive it while the mouse was over the UI around it.
> 
> I suppose this is also related to your other concern, since the input I'm using in the scene that's under the viewport is actually Area `input_event` signals (for mouse clicks and movement), which according to the docs is neither `_input` nor `_unhandled_input` and would normally be handled last.

I still can't get it to work.  I tried the workaround and v3.2 alpha (someone else had mentioned that it was working in it).  Perhaps I am not setting it up correctly?

**Comment by OlexiyKravchuk** [comment]
It's a little hack but it helped me.

```
func _input(ev) -> void:
	
	get_viewport().unhandled_input(ev)
```

**Comment by nikitavondel** [comment]
I still cannot get mouse_entered/exited signals to fire either inside a viewport of a viewportcontainer.

**Comment by wraitii** [comment]
Investigated this deeper today:
For reference, everything is working as designed, only the design is being difficult :p .
Assuming a viewport in a container, with object picking on and everything correctly setup.
What happens:
1. User clicks in the viewport somewhere.
2. the input is propagated upwards. Here lies the first blocker -> most likely a control will have mouse.filter 'stop' then, and things stop here.
3. If the event wasn't stopped or handled, the viewport unhandled_input will be called.
4. Only then does it create a deferred physics picking call.
5. The deferred physics picking call triggers the collision Objects 'InputEvent', sending signals.

This has several flaws:
- If you want to actually trigger the object picking, you generally need to manually trigger 'unhandled input' (as [here](https://github.com/godotengine/godot/issues/26181#issuecomment-471229783)), as there's basically always a node with a "stop" mouse filter.
- Because the physics picking pass is async, input cannot be blocked by picking.
- It also means that one cannot get a 'nothing was picked' input event easily.

So if you have a viewport with some items, and you'd want clicking somewhere on the viewport but _not_ on an item to trigger something, the only solution is to have another area behind those objects (from the camera POV) that gets picked. And when that happens, you can do things (possibly send an input event 'behind' for example).

I have to say, because object picking happens during physics processing, I can't see an easy fix to this (not a godot dev though, so perhaps you guys have an idea already).
Perhaps with object picking activated, viewports should 'hold' events for at least one physics-processing-frame, and then only bubble them upwards if nothing was picked. This would introduce some slight input lag, but it _would_ make the whole architecture much more predictable, ad you'd either get an object pick event, or a regular GUI input.

**Comment by konsvasi** [comment]
I've got a similar issue with the latest version 3.2.1.stable.official of Godot, I tried the hack with the `unhandled_input` and it worked but now the `mouse_exited` signal is triggered all the time, even when I'm within the `Area2D` the `mouse_exited` event is fired all the time.

**Comment by anders-s** [comment]
Same issue on 3.2. Connecting the `mouse_entered` signal from an `Area2D` to a script never fires that script function. Worked around the same way @nikitaab20 did. Could be my fault, I'm new to Godot.

**Comment by luisfmnunes** [comment]
Just faced the same problem, but instead of a viewport it was a Panel node that wasn't propagating the signal to an Area2D node BELOW (in the common parent tree), the unhandled_input trick solved it (indicating that it inherits the same problem from the viewport).

**Comment by Duehok** [comment]
Still broken. The workaround does not work. _input_event never fires for an Collision2D node in a viewport.
Use case example: A "map" renders to a viewport in a viewportContainer. The use of a viewport allows to scroll the map by moving the camera.

This bug (yes, it's a bug) makes it impossible to select a unit, building, city, etc... by clicking on it.

**Comment by db0** [comment]
I will add my voice that this seems like a bug or at the least like a very unintuitive issue. There's no reason why a Viewport would be handling input events for children nodes if the ViewportContainer has its mouse_filter set to ignore. It makes even less sense that it *does not* appear to be handling events going to control nodes, but only stuff going towards Area2D/CollisionShape2D

**Comment by kone9** [comment]
2020 and still the same problem
https://www.youtube.com/watch?v=gp-qCWfLgNU&feature=youtu.be
<iframe width="560" height="315" src="https://www.youtube.com/embed/gp-qCWfLgNU" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

**Comment by jackwilsdon** [comment]
Is this not a duplicate of #17326?

**Comment by jdferreira** [comment]
I've lost about two days with this issue and have finally been able to solve it. What finally clicked for me was wraitii's [comment](https://github.com/godotengine/godot/issues/26181#issuecomment-565819411), in particular:

- making sure that all upward nodes from the viewport had the `mouse_filter` set to "Pass";
- making sure that `physics_object_picking` on the viewport was set to `true`
- getting rid of the `_input` and `_unhandled_input` functions on the viewport parent (a `ViewportContainer`)
- adding the following to the parent
```
func _gui_input(event):
    $Viewport.unhandled_input(event)
```

I'm not entirely sure this solves the issue, it may not for others, but it did for me.
@kone9, would you be willing to test that may bullet points above are able to solve the issue for you?

This may not be a very frequent situation, since it requires handling viewports and mouse input, which I suspect is unusual  in game development with Godot. But it should be much more well documented regarding the solution. I'd be happy to contribute but I have no clue where to do it. If someone can point me to the right direction I'd write a few lines in the documentation about this.

**Comment by Calinou** [comment]
> This may not be a very frequent situation, since it requires handling viewports and mouse input, which I suspect is unusual in game development with Godot. But it should be much more well documented regarding the solution. I'd be happy to contribute but I have no clue where to do it. If someone can point me to the right direction I'd write a few lines in the documentation about this.

See [Contributing to the Class Reference with Git](https://docs.godotengine.org/en/latest/community/contributing/updating_the_class_reference.html) :slightly_smiling_face:

**Comment by mg979** [comment]
I also had the same issue and got it fixed thanks to https://github.com/godotengine/godot/issues/26181#issuecomment-565819411.

And what @jdferreira said in https://github.com/godotengine/godot/issues/26181#issuecomment-761903838 is also important.

My issue was in a 3d viewport. All this meant to me that as soon as you use a viewport, handling inputs in it becomes impossible if you don't know what's going on, that is if you don't read this issue. I don't have enough knowledge about this stuff, but it seems to me that a _viewport container_ should forward automatically unhandled inputs to a child viewport, what's the point of blocking them?

**Comment by DoubleF3lix** [comment]
I had the same issue with an Area3D node but none of the suggested workarounds have worked for me

**Comment by DoubleF3lix** [comment]
Okay, after about 3 months of struggling with this on and off, [jdferreira's fix](https://github.com/godotengine/godot/issues/26181#issuecomment-761903838) worked for me. Mainly, make sure "Object Picking" under "Physics" in your viewport is checked. I had the rest of the system setup, and now it works fine.

**Comment by jeremyz** [comment]
It's working fire with current head : 59b69500 if you tick Physics : Object Picking in the Viewport.

**Comment by gilzoide** [comment]
> but it seems to me that a viewport container should forward automatically unhandled inputs to a child viewport, what's the point of blocking them?

I totally agree. Inputs could even only be propagated when `Viewport.physics_object_picking` is true.

This may not be a usual situation in games, but I'd guess it is for desktop/web tools with mouse input.
I'm always seeing more and more tools made with Godot (which is awesome!).

**Comment by jeremyz** [comment]
> Use case example: A "map" renders to a viewport in a viewportContainer. The use of a viewport allows to scroll the map by moving the camera.

It found a work around, prioritizing units Area2d over map Area2D  [boardgame-map](https://github.com/jeremyz/godot-stuff/tree/master/boardgame-map)

**Comment by nikolamalesevic** [comment]
>     * adding the following to the parent

@jdferreira Can you please explain which node do you refer to as **parent**? The parent of the `Viewport` - `ViewportContainer`?
I tried following your approach, but to no avail - I still cannot make any mouse picking to work in my viewport (apart on gui elements).

**Comment by colorworlds** [comment]
Hello,I have a similar problem.
Then I set following setting:
```
ViewportContainer must have mouse_filter set to IGNORE
Viewport must have Physics : Object Picking ticked
```
It worked well on PC.
But When I export to Android, It failed again. Inputs become unresponsive.
I use v3.3.2.stable.offical on win7.
I think it may be caused by ignoring the touch screen event ? I'm not sure

**Comment by jamie-pate** [comment]
This seems to work without the workaround if you set `$ViewportContainer.mouse_filter = ignore`

ViewportContainer -> Viewport -> Area.input_event signals will fire.

Seems backwards though :D

**Comment by colorworlds** [comment]
邮件已查收~ ( *^o^* ) つロ 乾杯~

**Comment by Sauermann** [comment]
This is a duplicate of #17326.
